runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -11658,12 +11659,70 @@ J9::Power::CodeGenerator::inlineDirectCall(TR::Node *node, TR::Register *&result
             TR::Node *destOffset = node->getChild(2);
             TR::Node *len = node->getChild(3);
             TR::Node *byteValue = node->getChild(4);
-            dest = TR::Node::create(TR::aladd, 2, dest, destOffset);
 
-            TR::Node * copyMemNode = TR::Node::createWithSymRef(TR::arrayset, 3, 3, dest, len, byteValue, node->getSymbolReference());
-            copyMemNode->setByteCodeInfo(node->getByteCodeInfo());
+            // When using balanced GC policy with offheap allocation enabled, there are three possible cases:
+            //
+            // 1.) The object at dest is known to be a non-array object at compile time. In this scenario, the final destination address
+            //     can be calculated by simply adding dest and destOffset.
+            // 2.) The object at dest is known to be an array at compile time. In this scenario, we must load dataAddr from dest
+            //     to account for the way arrays are allocated when offheap is enabled. Once this adjustment is made, the final
+            //     destination address can be calculated by adding the adjusted dest and destOffset, similar to case (1)
+            // 3.) The type of the object at dest is unknown at compile time. In this scenario, a runtime arrayCHK must be generated to
+            //     determine whether dataAddr needs to be loaded before calculating the final destination address. Thus, dest
+            //     and destOffset will be passed to setmemoryEvaluator() separately so that both possible control flow paths are accounted
+            //     for in the generated assembly code.
+            //
+            // Note that if the default (gencon) GC policy is being used, dataAddr will never need to be loaded no matter
+            // the type of the object at dest, so all of the above cases will be treated like case (1).
+
+            bool arrayCheckNeeded, adjustmentNeeded;
+
+            if (dest->getSymbolReference() != NULL)
```

Comment:
given the new info in get/put/cas, i think you need to dance to the same tune here as well (checking Interface and/or parameter).